nix/ops/solid/builder.sh

```diff
@@ -15,18 +15,41 @@ cleanup () {
 
 trap cleanup EXIT
 
-# update pill strategy to ensure correct staging
+#  update pill strategy to ensure correct staging
 #
+
 herb ./pier -p hood -d "+hood/mount /=home="
 
-cp $ARVO/app/lens.hoon ./pier/home/app/
-cp $ARVO/lib/pill.hoon ./pier/home/lib/
-chmod -R u+rw ./pier/home/lib/
+#  update :lens, :dojo and dependencies
+#
+#    XX reduce this list
+#
+cp $ARVO/app/lens.hoon   ./pier/home/app/      2>/dev/null || true
+cp $ARVO/app/dojo.hoon   ./pier/home/app/      2>/dev/null || true
+cp $ARVO/lib/base64.hoon ./pier/home/lib/      2>/dev/null || true
+cp $ARVO/lib/server.hoon ./pier/home/lib/      2>/dev/null || true
+cp $ARVO/lib/sole.hoon   ./pier/home/lib/      2>/dev/null || true
+mkdir -p ./pier/home/mar/lens/
+cp $ARVO/mar/lens/*      ./pier/home/mar/lens/ 2>/dev/null || true
+
+cp $ARVO/sur/lens.hoon   ./pier/home/sur/      2>/dev/null || true
+cp $ARVO/sur/sole.hoon   ./pier/home/sur/      2>/dev/null || true
+
+#  update +solid and its dependencies
+#
+cp $ARVO/lib/pill.hoon   ./pier/home/lib/      2>/dev/null || true
+cp $ARVO/gen/solid.hoon  ./pier/home/gen/      2>/dev/null || true
+
+chmod -R u+rw ./pier/home/
 
 herb ./pier -p hood -d "+hood/commit %home"
 herb ./pier -p hood -d "+hood/unmount %home"
 
-# stage new desk for pill contents
+#  XX horrible hack to ensure the update is applied first
+#
+sleep 10
```

Comment:
@philipcmonk, without this `sleep`, the subsequent `:dojo` `+poke-lens-command` crashes with `%ap-slam-mismatch` when `sur/lens` changes. In other words, the now-asynchronous responses from clay no longer block herb/lens, so the poke arrives while we're in an inconsistent state. I'm curious if you have any ideas for how this might more-elegantly be handled.

Reply:
The right solution is to block on verifying that the thing is done.  For example, here I needed to block until the mount went through, so I waited until the directory existed: https://github.com/urbit/urbit/blob/1391ecafd39ce928592438aaba1baf5f59f0c75a/nix/ops/solid/builder.sh#L20-25

In this case, I guess we need to block on making sure the file has the correct contents, right?  We can retrieve the contents of a file through herb/scry, right?

Reply:
Do you need the files to be correct or the app?  Another option is to put unmount into the write queue (like mount is), so that it'll only run once the commit has finished.  But there's no guarantee that the app will have updated.

Reply:
The only way to scry with herb right now is through hoon evaluation in `:dojo`, and we can't poke dojo with anything until it's rebuilt per the update.

There's an implicit requirement that `:lens` and `:dojo` stay in sync with regard to the types they're using, but there doesn't seem to be a way to make that explicit. Perhaps herb should be able to scry directly without going through apps/pokes/marks (maybe through automatic %eyre support for scry urls).

Reply:
Another issue is that `%ap-slam-mismatch` on a `%poke` is just swallowed by gall -- it doesn't result in a `%coup` error to the poke-er. If it did, :lens could produces an HTTP 500 and herb could be made to retry.

Reply:
@jtobin, this `%ap-slam-mismatch` issue is something to track wrt %gall fixes.